Cli: correctly set up preset values in `PreparePreset()`

samhutchins opened the pull request:
**Description of Change:**

`PreparePreset` is used to override properties in the preset with values from the command line, and it has some logic for applying default values in the case where no defaults are available. I believe the intent is to take either the last specified bitrate or the last specified audio quality (with quality taking priority in ties) and apply it to all the uninitialised presets.

~Currently this is only done when a _quality_ is supplied on the command line, and not when bitrates are supplied. This patch ensures this happens when either a bitrate or a quality is supplied~

**Tested on:**

- [ ] Windows 10+  (via MinGW)
- [x] macOS 10.13+
- [x] Ubuntu Linux

samhutchins committed:
Cli: ensure last bitrate/quality is applied when either quality or bitrate supplied

samhutchins commented:
I've slept on this and had another look, and either I don't understand what this is meant to do (very possible) or it doesn't do what it's meant to do.

`PreparePreset` modifies preset audio behaviour if audio settings are given on the command line without also specifying explicit tracks on the command line: https://github.com/HandBrake/HandBrake/blob/dd2548223f8ed7ceb4c9c81a369970c4e5394000/test/test.c#L3895-L3906

It gets the `AudioList` out of the preset, an initialises an empty list if it's not there: https://github.com/HandBrake/HandBrake/blob/dd2548223f8ed7ceb4c9c81a369970c4e5394000/test/test.c#L3912-L3918

It works out how many items _should_ be in that list based on the maximum number of arguments passed to any of the audio settings: https://github.com/HandBrake/HandBrake/blob/dd2548223f8ed7ceb4c9c81a369970c4e5394000/test/test.c#L3919-L3929

All makes sense so far

Next it finds the last dict in the list (`last_audio_dict`): https://github.com/HandBrake/HandBrake/blob/dd2548223f8ed7ceb4c9c81a369970c4e5394000/test/test.c#L3933-L3941

At this point, `last_audio_dict` can only be null if the list is empty, which happens if the list in the preset is empty, or if it wasn't there and the above code initialised an empty one.

Here's a preset with an empty audio list: [empty_audio.json.zip](https://github.com/HandBrake/HandBrake/files/14954958/empty_audio.json.zip)

Next it duplicates the last item in the list as needed to make the length of the list match the count that was calculated earlier. This only happens if `last_audio_dict` is _not_ null, which means if the list in the preset is empty nothing happens here, and the list is still empty

https://github.com/HandBrake/HandBrake/blob/dd2548223f8ed7ceb4c9c81a369970c4e5394000/test/test.c#L3944-L3948

Next it updates codecs: https://github.com/HandBrake/HandBrake/blob/dd2548223f8ed7ceb4c9c81a369970c4e5394000/test/test.c#L3951.

Then it has this logic to apply the last specified codec to remaining entries in the list: https://github.com/HandBrake/HandBrake/blob/dd2548223f8ed7ceb4c9c81a369970c4e5394000/test/test.c#L3962-L3972

And this is where things don't make sense to me. This is only done if `last_audio_dict` is `NULL`, and if this is `NULL` that means the audio list is empty, which means that the `audio_dict` extracted on line 3968 _must_ be null, and the `hb_dict_set` on the following line does nothing. And I believe this will always be the case.

You can see this with [this preset](https://github.com/HandBrake/HandBrake/files/14954958/empty_audio.json.zip) and the following command line:

` HandBrakeCLI --preset-import-file empty_audio.json --aencoder mp3 --mixdown stereo,stereo,stereo --preset-export test`

You can see in the preset that's printed out that no audio tracks have been added to the preset.

galad87 commented:
Yes, that looks like it will be always NULL unless the audio list is not empty. I don't know what was the intention of that code, maybe it was supposed to add a new audio_dict if the AudioList was empty.

samhutchins committed:
Cli: correctly set up preset values

When there are more settings specified than items in the AudioList list, HandBrake needs
to add more items. In the case where the AudioList in the preset has items, we just
duplicate the last item and leave it at that. In the case there the AudioList is empty, we
want to use the last setting supplied on the command line in all the remaining items.

samhutchins commented:
I assume that the intent is to have slightly different behaviour depending on whether or not the source preset has any values in `AudioList`

If it does have values, it should just duplicate the last item in the list, and not override any values that weren't explicitly given on the command line

If it doesn't have values, it should add an empty item then use the last value on the command line in any remaining un-initialised items

galad87 commented:
I think it's ok, but let's wait a bit before merging if someone has some additional comments.

samhutchins commented:
👍🏻. I think @jstebbins was the last person to modify this code, in 972aef5ea17869b7a36c48ffd54d68e1c6d975b5

jstebbins commented on test/test.c:
```diff
@@ -3929,23 +3929,30 @@ static hb_dict_t * PreparePreset(const char *preset_name)
             if (audio_dict == NULL)
             {
-                break;
+                // More settings specified than preset currently supports.
```
This breaks all subsequent logic that uses last_audio_dict since the loop will continue setting last_audio_dict to the newly created dict. You can fix by adding `continue;` after `hb_vaule_array_append()`

jstebbins commented:
The intended behavior of this piece of code is to allow overriding (and adding to) any audio setting in the preset's audio encoder settings list. The preset's audio list is a list of encoder settings to be applied to each track that is selected via the audio track selection behavior settings in the preset. So this code is only applied when there are *no* explicit audio tracks specified on the command line.

It would help if you stated what exactly you are trying to fix.  What is the incorrect behavior that you are attempting to modify?

After re-reading the code (triggered by my attempt to understand your patch :wink: ), I do see unintended behavior if the preset has an empty audio list. It really should be initializing a new empty audio entry in this case that then gets dup'd for each override given on the command line.

I'm thinking something like:
```
// After finding last_audio_dict...

hb_dict_t * audio_dict_stub = NULL;
if (last_audio_dict == NULL)
{
    audio_dict_stub = last_audio_dict = hb_dict_init();
}

// After the reset of the preset override code...

if (audio_dict_stub != NULL)
{
    hb_dict_free(&audio_dict_stub);
}
```

jstebbins commented:
> Then it has this logic to apply the last specified codec to remaining entries in the list:
> 
> https://github.com/HandBrake/HandBrake/blob/dd2548223f8ed7ceb4c9c81a369970c4e5394000/test/test.c#L3962-L3972
> 
> And this is where things don't make sense to me. This is only done if `last_audio_dict` is `NULL`, and if this is `NULL` that means the audio list is empty, which means that the `audio_dict` extracted on line 3968 _must_ be null, and the `hb_dict_set` on the following line does nothing. And I believe this will always be the case.
> 
> You can see this with [this preset](https://github.com/HandBrake/HandBrake/files/14954958/empty_audio.json.zip) and the following command line:
> 
> ` HandBrakeCLI --preset-import-file empty_audio.json --aencoder mp3 --mixdown stereo,stereo,stereo --preset-export test`
> 
> You can see in the preset that's printed out that no audio tracks have been added to the preset.


This is definitely broken behavior and doesn't make sense as you say :grin: A lot went wrong there, oops.

last_audio_dict would have only been NULL if there were no list so there can only be a "last codec" if it were specified on the command line in `acodecs[]`. The case of no encoder specified in the preset audio list or command line is completely unhandled. And the logic for adding unspecified codecs is completely wrong. I don't know what was going through my mind at the time, but it's just wrong. 

When the list is empty, we need to apply a default codec (`hb_audio_encoder_get_default()`). When the list is not empty, the intent was to use the last encoder specified on the command line for the remaining items that do not have an encoder specified. It's kind of a command line shortcut, if you were going to repeat the same encoder (or other audio setting for that matter) for the remainder of the audio list, you only need to specify once.

jstebbins commented:
Just an FYI, I believe the only audio settings that *must* be set in the audio dict are `AudioEncoder` and either `AudioTrackQualityEnable` with `AudioTrackQuality` or `AudioBitrate`. libhb will handle defaults for the rest.

jstebbins commented:
Just thinking about the cleanest way to fix this. I'm thinking modify the code snippet I gave above to initialize `audio_dict_stub` with the last `acodecs[]`, and `abitrates[]`, or `aqualities[]` from the command line. If any of these are not given on the command line use appropriate default, e.g. `hb_audio_encoder_get_default()`, `hb_audio_bitrate_get_default()`, and `hb_audio_quality_get_default()`. Then for any of the other audio settings that are specified on the command line, use the last to init `audio_dict_stub`. e.g. if any `mixdowns[]` were specified, use the last to set a mixdown in `audio_dict_stub`. If not specified on the command line, no defaults are needed for these settings.

That code snippet with the above modification makes sure there is a complete audio list with appropriate defaults. So at that point I think all that is necessary is to process remaining command line overrides. No further need to consider `last_audio_dict` etc.

jstebbins commented:
Another thing I should point out. On the command line, it is allowed to signal "don't change this" by leaving an entry in the comma separated list of values empty. So something like `-E ffaac,lame,` indicates overriding 1st and 2nd encoder entries in the preset audio list, but leaves the 3rd (and all subsequent) as they were (note the trailing `,` which will result in `acodecs[2][0] == 0`. In this case, if the preset audio list has < 3 items, any items added due to overrides should be assigned the default audio encoder.

jstebbins commented:
I think maybe I should take a crack at fixing this. There are a lot of silly little details that are hard to communicate.

samhutchins commented:
> I think maybe I should take a crack at fixing this. There are a lot of silly little details that are hard to communicate.

Haha. That's fine by me, I did suspect there might be dragons around here. I'm also happy to go back and forth on it, either here or IRC. Whatever works best for you :-)

jstebbins commented:
Superseded by https://github.com/HandBrake/HandBrake/pull/5969
